File: server/src/com/thoughtworks/go/server/ui/plugins/PluginInfo.java
/*
 * Copyright 2016 ThoughtWorks, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.thoughtworks.go.server.ui.plugins;

import com.thoughtworks.go.plugin.api.info.PluginDescriptor;

public class PluginInfo {
    private final String id;
    private final String name;
    private final String version;
    private final String type;
    private final String displayName;
    private final PluggableInstanceSettings pluggableInstanceSettings;

    public PluginInfo(String id, String name, String version, String type, String displayName, PluggableInstanceSettings settings) {
        this.id = id;
        this.name = name;
        this.version = version;
        this.type = type;
        this.displayName = displayName;
        this.pluggableInstanceSettings = settings;
    }

    public PluginInfo(PluginDescriptor descriptor, String type, String displayName, PluggableInstanceSettings settings) {
        this(descriptor.id(), descriptor.about().name(), descriptor.about().version(), type, displayName, settings);
    }

    public PluginInfo(String id, String name, String version, String type, String displayName) {
        this(id, name, version, type, displayName, null);
    }

    public String getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public String getVersion() {
        return version;
    }

    public String getType() {
        return type;
    }

    public PluggableInstanceSettings getPluggableInstanceSettings() {
        return pluggableInstanceSettings;
    }

    public String getDisplayName() {
        return displayName;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        PluginInfo that = (PluginInfo) o;

        if (id != null ? !id.equals(that.id) : that.id != null) return false;
        if (name != null ? !name.equals(that.name) : that.name != null) return false;
        if (version != null ? !version.equals(that.version) : that.version != null) return false;
        if (type != null ? !type.equals(that.type) : that.type != null) return false;
        if (displayName != null ? !displayName.equals(that.displayName) : that.displayName != null) return false;
        return pluggableInstanceSettings != null ? pluggableInstanceSettings.equals(that.pluggableInstanceSettings) : that.pluggableInstanceSettings == null;
    }

    @Override
    public int hashCode() {
        int result = id != null ? id.hashCode() : 0;
        result = 31 * result + (name != null ? name.hashCode() : 0);
        result = 31 * result + (version != null ? version.hashCode() : 0);
        result = 31 * result + (type != null ? type.hashCode() : 0);
        result = 31 * result + (displayName != null ? displayName.hashCode() : 0);
        result = 31 * result + (pluggableInstanceSettings != null ? pluggableInstanceSettings.hashCode() : 0);
        return result;
    }
}

File: server/webapp/WEB-INF/rails/webpack/models/shared/plugin_infos_new/plugin_info.ts
/*
 * Copyright 2018 ThoughtWorks, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as _ from "lodash";
import {About} from "./about";
import {ExtensionType} from "./extension_type";
import {Extension} from "./extensions";

enum State {
  active  = "active",
  invalid = "invalid"
}

class Status {
  readonly state: State;
  readonly messages?: string[];

  constructor(state: State, messages?: string[]) {
    this.state    = state;
    this.messages = messages;
  }

  static fromJSON(data: any) {
    return new Status(data.state, data.messages);
  }
}

export class PluginInfo<T extends Extension> {
  readonly id: string;
  readonly about: About;
  readonly imageUrl: string;
  readonly status: Status;
  readonly pluginFileLocation: string;
  readonly bundledPlugin: boolean;
  readonly extensions: T[];

  constructor(id: string,
              about: About,
              imageUrl: string,
              status: Status,
              pluginFileLocation: string,
              bundledPlugin: boolean,
              extensions: T[]) {
    this.id                 = id;
    this.about              = about;
    this.imageUrl           = imageUrl;
    this.status             = status;
    this.pluginFileLocation = pluginFileLocation;
    this.bundledPlugin      = bundledPlugin;
    this.extensions         = extensions;
  }

  static fromJSON(data: any, links?: any) {
    const extensions: Extension[] = data.extensions.map((extension: any) => Extension.fromJSON(extension));
    const imageUrl                = (links && links.image && links.image.href) || "";
    return new PluginInfo(data.id, About.fromJSON(data.about), imageUrl, Status.fromJSON(data.status),
      data.plugin_file_location, data.bundled_plugin, extensions);
  }

  supportsPluginSettings(): boolean {
    return this.firstExtensionWithPluginSettings() !== undefined;
  }

  types(): ExtensionType[] {
    return this.extensions.map((ext: Extension) => ext.type);
  }

  extensionOfType(type: ExtensionType): T | undefined {
    return this.extensions.find((ext: T) => ext.type === type);
  }

  firstExtensionWithPluginSettings(): T | undefined {
    return this.extensions.find((ext) => ext.hasPluginSettings());
  }

  hasErrors(): boolean {
    return this.status.state === State.invalid;
  }

  getErrors(): string {
    if (this.hasErrors() && this.status.messages) {
      return _.join(this.status.messages, " ");
    }
    return "";
  }
}
